File: src/data/code/merge_sort/code.js
function merge(arr, start, mid, end) { // 11
	const leftArr = arr.slice(start, mid + 1); // 12
	const rightArr = arr.slice(mid + 1, end + 1); // 13

	let left = 0; 
	let right = 0; 
	let out = start; // 14

	while (left < leftArr.length && right < rightArr.length) { // 15
		if (leftArr[left] <= rightArr[right]) { // 16
			arr[out] = leftArr[left]; // 17
			left++; // 18
		} else { // 19
			arr[out] = rightArr[right]; // 20
			right++; // 21
		}
		out++; // 22
	}

	while (left < leftArr.length) { // 23
		arr[out] = leftArr[left]; // 24
		left++; // 25
		out++; // 26
	}
}

function sort(arr, start, end) { // 4
	if (start >= end) { // 5
		return; // 6
	}

	const mid = Math.floor((start + end) / 2); // 7

	sort(arr, start, mid); // 8
	sort(arr, mid + 1, end); // 9
	merge(arr, start, mid, end); // 10
}

function mergeSort(arr) { // 2
	sort(arr, 0, arr.length - 1); // 3
}

const arr = [ ]; // 0
mergeSort(arr); // 1
console.log(arr); // 27

File: src/data/code/merge_sort/code.py
def merge(arr, start, mid, end): # 11
    left_arr = arr[start:mid+1] # 12
    right_arr = arr[mid+1:end+1] # 13

    left = 0
    right = 0
    out = start # 14

    while left < len(left_arr) and right < len(right_arr): # 15
        if left_arr[left] <= right_arr[right]: # 16
            arr[out] = left_arr[left] # 17
            left += 1 # 18
        else: # 19
            arr[out] = right_arr[right] # 20
            right += 1 # 21
        out += 1 # 22

    while left < len(left_arr): # 23
        arr[out] = left_arr[left] # 24
        left += 1 # 25
        out += 1 # 26

def sort(arr, start, end): # 4
    if start >= end: # 5
        return # 6

    mid = (start + end) // 2 # 7

    sort(arr, start, mid) # 8
    sort(arr, mid+1, end) # 9
    merge(arr, start, mid, end) # 10

def merge_sort(arr): # 2
    sort(arr, 0, len(arr)-1) # 3

arr = [ ] # 0
merge_sort(arr) # 1
print(arr) # 27

